ui/src/auth/AuthController.js

```diff
@@ -67,7 +67,14 @@ export default class AuthController {
     if (user) {
       const expires = new Date(user.expires);
       const now = new Date();
-      const timeout = Math.max(0, expires.getTime() - now.getTime());
+      let timeout = Math.max(0, expires.getTime() - now.getTime());
+
+      // if the timeout is in the future, apply up to a few minutes to it
+      // randomly.  This avoids multiple tabs all trying to renew at the
+      // same time.
```

Comment:
In https://github.com/taskcluster/taskcluster-tools/blob/17302c900904b892bc66f359c683e855beccb9a3/src/auth/AuthController.js#L47, we renew after "renewAfter", and that's for renewing the auth0 access_token (calling `auth0Renew`), rather than renewing TC credentials at expiration.  So that code is "cheating" a little since it knows that the renewal time is just a suggestion, not an expiration time.

In this case, the credentials actually *expire* at `expires`, so waiting up to five minutes after that will mean up to five minutes of failing API calls -- no bueno.

I bet just subtracting a random time instead of adding it (while ensuring we don't pick a time in the past) would fix this.

Reply:
Makes a lot of sense.